url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/leetcode/TreeFromInPreOrder.java

```diff
@@ -0,0 +1,64 @@
+package leetcode;
+
+import common.TreeNode;
+
+/**
+ * Created by sobercheg on 12/22/13.
+ */
+public class TreeFromInPreOrder {
+    /*
+         ______7______
+       /              \
+    __10__          ___2
+   /      \        /
+   4       3      _8
+            \    /
+             1  11
+
+
+   preorder = {7,10,4,3,1,2,8,11}
+   inorder = {4,10,3,1,7,11,8,2}
+     */
+
+    class MutableInt {
+        int i;
+
+        MutableInt(int i) {
+            this.i = i;
+        }
+    }
+
+    public TreeNode buildInorderPreorder(int[] pre, int[] in) {
+        return buildInorderPreorder(pre, new MutableInt(0), pre.length - 1, in, 0, in.length - 1);
+    }
+
+    public TreeNode buildInorderPreorder(int[] pre, MutableInt preStart, int preEnd, int[] in, int inStart, int inEnd) {
+        if (preStart.i > preEnd || inStart > inEnd) return null;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;preStart&#39; (index: 1 | type: MutableInt). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;preEnd&#39;, &#39;in&#39;, &#39;pre&#39;, &#39;inEnd&#39;, &#39;inStart&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).